Add all-gather/reduce-scatter coalescee for FSDP.

jeffhataws opened the pull request:
This PR adds all-gather and reduce-scatter coalescence support and use that in FSDP. Also it enables using reduce-scatter's scale param in FSDP. This PR is to be used in conjunction with https://github.com/openxla/xla/pull/5740 .

This is a revival of https://github.com/pytorch/xla/pull/4145 . Will need to address the comments.

JackCaoG commented:
@alanwaketan can you take a look?

alanwaketan commented:
@jeffhataws Can you double check the CI failures?

jeffhataws commented:
> @jeffhataws Can you double check the CI failures?

Thanks. Since it depends on https://github.com/openxla/xla/pull/5740 I will need to take care of merging that first. So let's leave this open for now.

jeffhataws commented:
@alanwaketan what's the best way to check this against the openxla with merged https://github.com/openxla/xla/pull/5740?

alanwaketan commented:
> @alanwaketan what's the best way to check this against the openxla with merged [openxla/xla#5740](https://github.com/openxla/xla/pull/5740)?

I'm working on a pin update. Will loop you in once that PR is up.

jeffhataws commented:
> > @alanwaketan what's the best way to check this against the openxla with merged [openxla/xla#5740](https://github.com/openxla/xla/pull/5740)?
> 
> I'm working on a pin update. Will loop you in once that PR is up.

@alanwaketan just want to check how things are going with the pin update. Also, how do I make the automatic checks to run on this PR?

jeffhataws commented:
@alanwaketan @JackCaoG  will you help ensure this is in 2.2?

JackCaoG commented:
@jeffhataws can you rebase this pr? Then we can start reviewing it. Thanks!

alanwaketan commented:
> > @alanwaketan what's the best way to check this against the openxla with merged [openxla/xla#5740](https://github.com/openxla/xla/pull/5740)?
> 
> I'm working on a pin update. Will loop you in once that PR is up.

The pin update is completed. So, please rebase and then I will take a look.

jeffhataws commented:
> > > @alanwaketan what's the best way to check this against the openxla with merged [openxla/xla#5740](https://github.com/openxla/xla/pull/5740)?
> > 
> > 
> > I'm working on a pin update. Will loop you in once that PR is up.
> 
> The pin update is completed. So, please rebase and then I will take a look.

Thanks [alanwaketan](https://github.com/alanwaketan). I have rebased. Please take a look.

jeffhataws commented:
@alanwaketan alanwaketan @JackCaoG   I am not sure how to reproduce/debug the above errors.

`test/run_tests.sh` passes for me on CPU machine.

When I run `test/cpp/run_tests.sh ` I get a hang:

```
Running all cpp test...
+ '[' /tmp/pytorch_cpp_test.log '!=' '' ']'
+ bazel test --config=tpu //torch_xla/csrc/runtime:all //test/cpp:all --test_timeout 1000
(hangs)
```

jeffhataws commented:
> @alanwaketan @JackCaoG I am not sure how to reproduce/debug the above errors.
> 
> `test/run_tests.sh` passes for me on CPU machine with `export PJRT_DEVICE=CPU`.
> 
> When I run `test/cpp/run_tests.sh ` with `export PJRT_DEVICE=CPU` I get a hang:
> 
> ```
> Running all cpp test...
> + '[' /tmp/pytorch_cpp_test.log '!=' '' ']'
> + bazel test --config=tpu //torch_xla/csrc/runtime:all //test/cpp:all --test_timeout 1000
> (hangs)
> ```

It wasn't hanging. After a while, it shows:

```
+ bazel test --config=tpu //torch_xla/csrc/runtime:all //test/cpp:all --test_timeout 1000
^[[A^[[A^[[A^[[A//test/cpp:test_aten_xla_tensor_1                                        PASSED in 23.4s
//test/cpp:test_aten_xla_tensor_2                                        PASSED in 18.5s
//test/cpp:test_aten_xla_tensor_3                                        PASSED in 44.9s
//test/cpp:test_aten_xla_tensor_4                                        PASSED in 15.4s
//test/cpp:test_aten_xla_tensor_5                                        PASSED in 85.5s
//test/cpp:test_aten_xla_tensor_6                                        PASSED in 44.5s
//test/cpp:test_ir                                                       PASSED in 0.5s
//test/cpp:test_lazy                                                     PASSED in 0.5s
//test/cpp:test_replication                                              PASSED in 0.6s
//test/cpp:test_tensor                                                   PASSED in 101.9s
//test/cpp:test_xla_sharding                                             PASSED in 0.5s
//torch_xla/csrc/runtime:cache_test                                      PASSED in 0.0s
//torch_xla/csrc/runtime:pjrt_computation_client_test                    PASSED in 0.7s
//torch_xla/csrc/runtime:sys_util_test                                   PASSED in 0.0s
//torch_xla/csrc/runtime:util_test                                       PASSED in 0.1s
//torch_xla/csrc/runtime:xla_util_test                                   PASSED in 0.4s

Executed 16 out of 16 tests: 16 tests pass.
There were tests whose specified size is too big. Use the --test_verbose_timeout_warnings command line option to see which ones these are.

```

JackCaoG commented:
You didn't add any cpp test and our cpp test doesn't test distributed so that should be fine? I do see a bunch of python test failing in the CI with
```
ImportError: /opt/conda/lib/python3.8/site-packages/torch_xla-2.2.0+git08075a8-py3.8-linux-x86_64.egg/_XLAC.cpython-38-x86_64-linux-gnu.so: undefined symbol: _ZN9torch_xla14tensor_methods10all_gatherERKN3c1013intrusive_ptrINS_9XLATensorENS1_6detail34intrusive_target_default_null_typeIS3_EEEEllSt6vectorISA_IlSaIlEESaISC_EEb
```
in https://github.com/pytorch/xla/actions/runs/6896526398/job/18766614593?pr=5624


If I have to guess it is that the build pass but when we try to import the XLAC, we find that there are some functions that only has the header but missing in the cpp file? this one seems to be `tensor_methods::all_gather`. If you try to build this locally and just do a `import torch_xla` you should see the same failure.

JackCaoG commented:
@alanwaketan is out next week, I will work with you to try to land this change before branch cut(or we will cherry-pick).

jeffhataws commented:
> You didn't add any cpp test and our cpp test doesn't test distributed so that should be fine? I do see a bunch of python test failing in the CI with
> 
> ```
> ImportError: /opt/conda/lib/python3.8/site-packages/torch_xla-2.2.0+git08075a8-py3.8-linux-x86_64.egg/_XLAC.cpython-38-x86_64-linux-gnu.so: undefined symbol: _ZN9torch_xla14tensor_methods10all_gatherERKN3c1013intrusive_ptrINS_9XLATensorENS1_6detail34intrusive_target_default_null_typeIS3_EEEEllSt6vectorISA_IlSaIlEESaISC_EEb
> ```
> 
> in https://github.com/pytorch/xla/actions/runs/6896526398/job/18766614593?pr=5624
> 
> If I have to guess it is that the build pass but when we try to import the XLAC, we find that there are some functions that only has the header but missing in the cpp file? this one seems to be `tensor_methods::all_gather`. If you try to build this locally and just do a `import torch_xla` you should see the same failure.

Fixed. Removed the previous all_gather method by mistake.

jeffhataws commented:
Fixed the lint error.

JackCaoG commented:
@jeffhataws there seems to be some python lint error as well, check https://github.com/pytorch/xla/actions/runs/6916906293/job/18864367391?pr=5624.

jeffhataws commented:
> @jeffhataws there seems to be some python lint error as well, check https://github.com/pytorch/xla/actions/runs/6916906293/job/18864367391?pr=5624.

Committed the lint fixs.

JackCaoG commented:
Let me just grant you the write access so your ci will autorun

JackCaoG commented:
OK now after you push CI should auto run, feel free to fix things. I will try to take a look at this pr today from high level.

jeffhataws commented:
> OK now after you push CI should auto run, feel free to fix things. I will try to take a look at this pr today from high level.

I had to switch to using yapf==0.30. The latest should work now.

jeffhataws commented:
I am able to reproduce locally now. Will debug.

JackCaoG commented:
for cpu `/tmp/pytorch/xla/test/pjrt/test_ddp.py` failed. For gpu multiple test failed one of them is `test_train_mp_imagenet_fsdp ` because of `RecursionError: maximum recursion depth exceeded while calling a Python object`.

JackCaoG commented:
We can keep trying to fix test errors, but I am not sure if it will be eaiser to split all-gather and reduce-scatter into separate prs. @jeffhataws I take that this pr works in your local forks, not sure what's the difference. Test failures seem real.

jeffhataws commented on torch_xla/core/xla_model.py:
```diff
@@ -582,6 +585,31 @@ def all_gather(value, dim=0, groups=None, output=None, pin_layout=True):
                                            [], pin_layout)
   return result
```
Should be removed? @jeffhataws

JackCaoG replied:
yea, if you want below code to be executed. seems like otherwise it will always return here.

jeffhataws commented on torch_xla/core/xla_model.py:
```diff
@@ -582,6 +585,31 @@ def all_gather(value, dim=0, groups=None, output=None, pin_layout=True):
+      torch_xla._XLAC._set_all_reduce_token(devctx.device, new_token)
+      return output
+
+    result = torch_xla._XLAC._xla_all_gather(value, token, dim, shard_count,
```
the new version has no token in port list, and return is just "return result"

jeffhataws commented on test/test_torch_distributed_xla_backend.py:
```diff
@@ -87,6 +87,26 @@ def test_allgather(self):
+    device = xm.xla_device()
+    tensor = torch.arange(2, device=device) + 1 + 2 * dist.get_rank()
+    tensor2 = torch.arange(5, device=device) + 1 + 2 * dist.get_rank()
+    pg_xla = get_process_group_xla(rank=3, size=8)
```
Updated get_process_group_xla:

from datetime import timedelta

```
def get_process_group_xla(rank, size):
  pg_xla_creator = dist.Backend._plugins['XLA'].creator_fn 
  pg_xla = pg_xla_creator(
      prefix_store=None, rank=rank, size=size, timeout=timedelta(minutes=1))
  return pg_xla
```

jeffhataws commented:
> We can keep trying to fix test errors, but I am not sure if it will be eaiser to split all-gather and reduce-scatter into separate prs. @jeffhataws I take that this pr works in your local forks, not sure what's the difference. Test failures seem real.

Yeah the code was ported from an old version of torch/xla so there were some merge errors. Plus the final version of openxla change https://github.com/openxla/xla/pull/5740 doesn't have token support, so I need to make the corresponding change here.

jeffhataws commented:
One of the CPU workflows failed with this:

```
[25,304 / 29,344] Compiling xla/mlir_hlo/mhlo/transforms/legalize_to_linalg/legalize_to_linalg.cc; 33s local ... (48 actions, 47 running)
[28,230 / 30,817] Compiling xla/service/gpu/cub_sort_kernel.cu.cc; 31s local ... (48 actions, 47 running)
[29,163 / 30,817] Compiling xla/service/algebraic_simplifier.cc; 29s local ... (48 actions, 47 running)
/home/ec2-user/actions-runner/_work/_temp/c18e5d69-957a-4e96-8ea6-714d1254694a.sh: line 1: 27242 Killed                  docker exec --privileged -u jenkins "${pid}" bash -c '.circleci/test.sh'
Error: Process completed with exit code 137.
```

JackCaoG commented:
hmm, seems like vm oom when building pytorch/xla...

JackCaoG commented:
Let's ignore the CPU failure and focus on GPU for now. GPU failures seems real.

jeffhataws commented:
All tests passing with Reduce-Scatter change separated out in https://github.com/pytorch/xla/pull/5938 .

jeffhataws commented on torch_xla/csrc/cross_replica_reduces.cpp:
```diff
@@ -210,30 +210,41 @@ AllToAllResult BuildAllToAll(xla::XlaOp input, xla::XlaOp token,
+          xla::AllGather(xla::Tuple(inputs[0].builder(), type_ctx.second.ops),
+                         dim, shard_count, cc_groups);
+    }
+    if (type_ctx.second.indices.size() > 1) {
```
Maybe better to use `if (ShapeHelper::ShapeOfXlaOp(all_gather_result).rank() == 0) {` here.

JackCaoG commented on test/test_torch_distributed_xla_backend.py:
```diff
@@ -87,6 +96,27 @@ def test_allgather(self):
+    hlo = torch_xla._XLAC._get_xla_tensors_hlo(output_tensors)
+    hlo_matches(hlo, all_gather_pattern)
+    # purge all computations attached the device.
+    xm.mark_step()
```
do we need this through? I thought in this test we never really executed any computation, world size and rank are all fake so maybe we shouldn;t try to execute it.

JackCaoG commented on torch_xla/core/xla_model.py:
```diff
@@ -570,17 +570,30 @@ def all_gather(value, dim=0, groups=None, output=None, pin_layout=True):
+    return result
+
+  # Now the input should be a list of Tensors.
+  if not isinstance(value, list) or any(
```
can we use `elif` here something like
```
elif isinstance(value, list) and all(isinstance(v, torch.Tensor) for v in value):
  # do your thing here
else:
  raise TypeError
```

This way it is easier to follow the conditions.

jeffhataws replied:
fixed in https://github.com/pytorch/xla/pull/5624

JackCaoG commented on torch_xla/csrc/cross_replica_reduces.cpp:
```diff
@@ -210,30 +210,41 @@ AllToAllResult BuildAllToAll(xla::XlaOp input, xla::XlaOp token,
+    xla::XlaOp all_gather_result;
+    if (pin_layout) {
+      all_gather_result = xla::AllGather(
+          xla::Tuple(inputs[0].builder(), type_ctx.second.ops), dim,
```
I think this means even if there is single element in the all-gather, we will wrap it inside the tuple.. I need to check with xla teams whether this has any speed implications.

JackCaoG commented on torch_xla/csrc/cross_replica_reduces.cpp:
```diff
@@ -210,30 +210,41 @@ AllToAllResult BuildAllToAll(xla::XlaOp input, xla::XlaOp token,
+      }
+    } else {
+      result[0] = all_gather_result;
+    }
```
I don't see how token is being used here, in all_reduce, we manually append it at the end of `ops`
```
  for (auto& type_ctx : redux.contexts) {
    xla::XlaOp token_op = MaybeConvertTo(chained_token, type_ctx.first);
    type_ctx.second.ops.push_back(token_op);
    type_ctx.second.operand_shapes.push_back(
        ShapeHelper::ShapeOfXlaOp(token_op));
```

JackCaoG replied:
I saw the `GetOperandList` below, but I think this does not gurante when you have multiple types, each types has a token.

jeffhataws replied:
Fix the usage of token. 

For multiple types case, I think for now we should ensure same type in the list.

JackCaoG commented on torch_xla/csrc/init_python_bindings.cpp:
```diff
@@ -1152,6 +1171,27 @@ void InitXlaModuleBindings(py::module m) {
+           int64_t shard_count, const py::list& groups, bool pin_layout) {
+          std::vector<std::vector<int64_t>> replica_groups =
+              CreateReduceGroups(groups);
+          std::vector<at::Tensor> result;
```
nit `results`

jeffhataws replied:
fixed in https://github.com/pytorch/xla/pull/5624

JackCaoG commented:
oh I know what's going on. Can you create a branch from pytorch/xla directly instead of creating from a fork. Fork can not used our cache so compilation will take much longer and easier to fail. I already give you write access to the project so you should be able to create a new branch from our repo directly.

jeffhataws commented:
> oh I know what's going on. Can you create a branch from pytorch/xla directly instead of creating from a fork. Fork can not used our cache so compilation will take much longer and easier to fail. I already give you write access to the project so you should be able to create a new branch from our repo directly.

Thanks. Changed to PR from a branch on pytorch/xla https://github.com/pytorch/xla/pull/5950 .
